spiffe/id.go

```diff
@@ -0,0 +1,242 @@
+	}
+
+	// trust domain validation
+	if options.trustDomainRequired {
+		if options.trustDomain == "" {
+			return errors.New("trust domain to validate against cannot be empty")
+		}
+		if id.Host != options.trustDomain {
+			return fmt.Errorf("%q does not belong to trust domain %q", id, options.trustDomain)
+		}
+	}
+
+	// id type validation
+	switch options.idType {
+	case anyId:
+	case trustDomainId:
+		if id.Path != "" {
+			return validationError("path is not empty")
+		}
+	case workloadId:
+		if id.Path == "" {
+			return validationError("path is empty")
+		}
+	default:
+		return validationError("internal error: unhandled id type %v", options.idType)
+	}
+
+	return nil
+}
+
+// ParseID parses the SPIFFE ID and makes sure it is valid according to
+// the specified validation mode.
+func ParseID(spiffeID string, mode ValidationMode) (*url.URL, error) {
+	u, err := url.Parse(spiffeID)
+	if err != nil {
+		return nil, fmt.Errorf("invalid SPIFFE ID: %v", err)
+	}
+
+	if err := ValidateIDURL(u, mode); err != nil {
+		return nil, err
+	}
+
+	return normalizeURL(u), nil
+}
+
+// ValidationMode is used to validate the SPIFFE ID being parsed/validated
+// should have.
+type ValidationMode interface {
+	validationOptions() validationOptions
+}
+
+type validationOptions struct {
+	trustDomain         string
+	trustDomainRequired bool
+	idType              idType
+}
+
+type validationMode struct {
+	options validationOptions
+}
+
+func (m validationMode) validationOptions() validationOptions {
+	return m.options
+}
+
+// Allows any well-formed SPIFFE ID
+func AllowAny() ValidationMode {
+	return validationMode{}
+}
+
+// Allows a well-formed SPIFFE ID for the specific trust domain (e.g. spiffe://domain.test/workload)
+func AllowTrustDomain(trustDomain string) ValidationMode {
+	return validationMode{
+		options: validationOptions{
+			trustDomain:         trustDomain,
+			trustDomainRequired: true,
+			idType:              trustDomainId,
+		},
+	}
+}
+
+// Allows a well-formed SPIFFE ID for a workload belonging to a specific trust domain (e.g. spiffe://domain.test/workload)
+func AllowTrustDomainWorkload(trustDomain string) ValidationMode {
+	return validationMode{
+		options: validationOptions{
+			trustDomain:         trustDomain,
+			trustDomainRequired: true,
+			idType:              workloadId,
+		},
+	}
+}
+
+// Allows a well-formed SPIFFE ID for any trust domain (e.g. spiffe://domain.test).
+func AllowAnyTrustDomain() ValidationMode {
+	return validationMode{
+		options: validationOptions{
+			idType: trustDomainId,
+		},
+	}
+}
+
+// Allows a well-formed SPIFFE ID for a workload belonging to any trust domain (e.g. spiffe://domain.test).
+func AllowAnyTrustDomainWorkload() ValidationMode {
+	return validationMode{
+		options: validationOptions{
+			idType: workloadId,
+		},
+	}
+}
+
+// NormalizeID normalizes the SPIFFE ID so it can be directly compared
+// for equality.
+func NormalizeID(id string, mode ValidationMode) (string, error) {
```

Comment:
What kind of stuff would we need to be normalizing here? I assume something that the url lib is doing for us?

Reply:
It lower cases scheme and host portions. I don't know if you recall the conversation around when this went into SPIRE.... If we don't want this in go-spiffe, I can rip it out.

Reply:
Yes I remember, I thought it was something specific to what we were doing in SPIRE but this makes more sense :) perhaps quickly update the comment to say this? The way it stands now led me to question why they aren't directly comparable